contrib/babelfishpg_tsql/src/catalog.c

```diff
@@ -6287,3 +6288,49 @@ get_tvp_typename_typeschemaname(char *proc_name, char *target_arg_name, char **t
 	if(!xactStarted)
 		CommitTransactionCommand();
 }
+
+/*
+ * Checks if the given member is 'bbf_role_admin' role and has the privilege 
+ * to grant a specified role through direct membership. This privilege is 
+ * determined by the presence of the 'admin_option' flag, which allows the 
+ * member to grant the role to others. 
+ * 
+ * @return true if the member is 'bbf_role_admin' and also has the admin option 
+ *         to grant the specified role; otherwise, returns false.
+ */
+bool
+bbf_check_member_has_direct_priv_to_grant_role(Oid member, Oid role)
+{
+	CatCList   *memlist;
+
+	/* 
+	 * TSQL specific behavior and member must be bbf_role_admin.
+	 * If not return false. 
+	 */
+	if (sql_dialect != SQL_DIALECT_TSQL || !IS_TDS_CONN()
+		|| member !=  get_bbf_role_admin_oid())
+		return false;
+
+	/* 
+	 * Find all existing tuples for given role
+	 * whose member is bbf_role_admin
+	 */
+	memlist = SearchSysCacheList2(AUTHMEMROLEMEM,
```

Comment:
we are checking two things
1) bbf admin is member of "role"
2) membership is with admin option ON

do we have a use case where this not true for bbf_admin and bbf_roles? if not, it might be sufficient to check for role being bbf_role (assuming that check is very fast to do)

Reply:
No, for every bbf_roles, these conditions will be true. I checked `is_babelfish_role` API is costlier than the current approach since it uses `roles_is_member_of` API internally to find all memberships.